src/libcollections/str.rs

```diff
@@ -705,7 +707,7 @@ pub trait StrExt: Index<RangeFull, Output = str> {
     /// ```
     #[unstable(feature = "collections",
                reason = "might have its iterator type changed")]
-    fn match_indices<'a>(&'a self, pat: &'a str) -> MatchIndices<'a> {
+    fn match_indices<'a, P: Pattern<'a>>(&'a self, pat: P) -> MatchIndices<'a, P> {
```

Comment:
There is still the question about how it should behave though - right now it returns `(usize, usize)`, but it might be more consistent with `char_indices()` and more useful in regard to arbitrary patterns to return `(usize, &str)`